README
README - Documentum Perl 5 Extension
====================================

This is the Documentum API interface to Perl.

These modules have been built and tested with EDMS versions 3.1.5, 3.1.6, 98,
4.0.2, 4.1, 4.2, 5.2, 5.2.5 SP2, 5.3 SP1.  They should also work with other versions,
as the API hasn't changed much in a long time.

* IMPORTANT NOTE #1*
The Makefile.pl is primarily designed to run in a Microsoft Windows
environment.  However, the script does contain configuration information
for Solaris, AIX, and HP-UX environments.  These portions of the Makefile.pl
script have been graciously donated by users and are mostly untested since I
don't have facilities to test them.  Therefore, if you are running in a
UNIX environment, you might need to tweak the Makefile.pl script.

* IMPORTANT NOTE #2*
This module has changed in one significant way from 1.01:  dmAPIInit()
and dmAPIDeinit() are now called implicitly within the module.  As such
your client programs do not need to call this.  Especially careful
client code which may check the return status of these calls will
complain or fail, depending on what you're doing with the results.

To build this extension, unpack the distribution, and edit Makefile.PL
if necessary.  Then do the typical Perl module building stuff:

	perl Makefile.PL
	make
	make test
	make install

The test.pl script will prompt you for the information it needs, and then
will connect to your docbase.  It will generate a new document (dm_document)
named 'Perl Module Test' and link it to the /Temp cabinet.  The test script
will also create a new object type named 'my_document', create an instance
of one named 'Perl Module Tools Test Doc' and link it to a new folder at
/Temp/Db-Documentum-Test'.  Finally, the script copies and moves the 'Perl
Module Tools Test Doc' among several folders and then deletes the whole mess.
If you're nervous about the test script doing evil things to your production
docbase, examine the code, point it at a test docbase, or write your own test
script.

To check the version of this module from the command line, use:
	>perl -MDb::Documentum -e Db::Documentum::version


Documentum::Tools
=================

This is a set of helper functions that simplify a couple of common
tasks when writing a Documentum client application.

dm_Connect
----------

	Encapsulates the process of connecting to a docbase.

	Arguments:  docbase, username, and password are required.  Two
		extra args are optional, and your dm_check_password program
		(or equivalent) should be configured to do the right thing
		with them.  See the etc/dm_check_password.pl program included
		with this distribution for an example that uses Kerberos v4
		service tickets to authenticate to Documentum.

	Returns:  the session identifer on success.  Nothing on failure.

dm_LastError
------------

	Generic interface for evaluating the error log for the current
	session.

	Arguments: session identifier, error level, and number of entries
		to report, all optional.  Session id defaults to 'apisession'
		for handling connect() failures.  Error level defaults to 3, which
		is a good choice in most situations, and number of entries
		defaults to 'all'.

	Returns:  character string containing the error data, suitable for
		printing to the user.

dm_CreateObject
---------------

    NOTE: The %ATTRS hash is no longer a HoL.  This is a change from v1.2

	Generic wrapper to create new instances of object types and optionally
	populate attributes.  Note: this function does NOT perform a 'save'.
	Repeating attributes are declared as a scalar with delimiters between
	repeating values.  See below.  The default delimiter is set at:
	$Db::Documentum::Tools::Delimiter

	Arguments: Documentum object type, required.  Hash of attributes for object,
	optional.  For example:

	    $delim = $Db::Documentum::Tools::Delimiter;
		%ATTRS = (object_name =>  'test_doc2',
               	  title       =>  'My Test Doc 2',
               	  authors     =>  'Scott 1'.$delim.'Scott 2',
              	  keywords    =>  'Scott'.$delim.'Test'.$delim.$DocVar.$delim.'2',
               	  r_version_label => 'TEST');

   		$doc_id = dm_CreateObject ("dm_document",%ATTRS);

	Returns: scalar containing the r_object_id of the new object.

dm_CreateType
-------------

	Generic wrapper to create new object type in the docbase.

	Arguments: Object name and Documentum superclass name, required.  Hash
	of field definitions, optional.  For example:

		%field_defs = (cat_id    => 'char(16)',
			           loc       => 'char(64)',
                       editions  => 'char(6) REPEATING');

     		$api_stat = dm_CreateType ("my_document","dm_document",%field_defs);

	Returns: TRUE on success, FALSE on failure.

dm_CreatePath
-------------

	Generic wrapper to create new folder heirarchies in Docbase.

	For example,
	$dir_id = dm_CreatePath ('/Temp/Testing/Unit-1');

	Returns: r_object_id of folder or undef.

dm_LocateServer
---------------

	Locates the currently-active server for a particular docbase.

	This is only really necessary since to obtain a Kerberos
	service ticket, you must know the hostname of the machine you're
	connecting to.  The Docbroker hides this layer from the user
	by default, and this routine exposes it.  This only gets
	called by dm_KrbConnect() in my implementation.

    Example:  $server = dm_LocateServer($docbase);
	Returns:  Hostname for docbase on success.  Nothing on failure.

dm_Copy
-------
    Copies an object and optionally moves it to a new location specified by
    $to_folder.  If $to_folder is undefined, the copy will reside with the
    original.  Do not copy folders and cabinets with this routine.  Returns
    the r_object_id of copy.

    Example:  $object_id = dm_Copy($orig_obj_id, $to_folder)

    Note:  $to_folder can be the r_object_id of the destination folder,
         or the full path (e.g., "/Temp/2004/Q1/Jan/15").  If the path
         does not exist, it is created by calling dm_CreatePath.

dm_Move
-------
    Moves an object from the location specified by $from_folder, to the
    location specified by $to_folder.  If $from_folder is undefined, unlink
    from all locations.  Returns true for success, false for failure.

    Example:  $rv = dm_Move($object_id, $to_folder, $from_folder)

    Note:  $to_folder can be the r_object_id of the destination folder,
         or the full path (e.g., "/Temp/2004/Q1/Jan/15").  If the path
         does not exist, it is created by calling dm_CreatePath.

    Note:  $from_folder can be the r_object_id of the current folder or its
         full path (e.g., "/Temp/2004/Q2/Jun/1").

dm_Delete
---------
    Deletes the object specified by $obj_id.  If the optional $all argument is
    set to true, will delete all versions of the object specified by $obj_id.
    By default, $all is false.  If the object specified by $obj_id is a cabinet
    or folder, this sub will execute a deep delete and remove all versions of
    all objects contained in the cabinet or folder.  Returns true on success,
    false on failure.

    Example:  $rv = dm_Delete($obj_id, $all)

dm_KrbConnect
-------------
    !! EXPERIMENTAL !!
	Requires Krb4.pm from CPAN.  You'll want to comment out the
	'use Krb4'; statement in the header of the library if you
	don't plan on using this.

	Connects to Documentum and authenticates the user using a Kerberos
	v4 service ticket.  This is fairly tricky stuff.  The library does a
	bunch of Kerberos calls to obtain all of the necessary info to
	build the service ticket.  The ticket is then encoded, along with
	a nonce value used to prevent replay attacks, and sent to the
	server.  A compatible dm_check_password program should be installed
	on the server to decode and decrypt the session info.

	You need to be fairly Kerberos-savvy to make this work.  The code is
fairly-well commented, but it does assume you have a working
	K4 KDC installed, and a service ticket entry in your server machine's
	srvtab.  I did my testing with the default rcmd.hostname@REALM ticket,
	but now use documentum.hostname@REALM (which is how this code is
	configured).  You'll need to configure the service in both the
	client library code as well as the dm_check_password.pl program if
	you want to change it.

	The big problem with this code is that there is no good way to
	map Kerberos problems to Documentum connect() error codes, and
	reporting reporting Kerberos messages back to the user is not
	possible given the interface between the authentication program
	and the user, which is simply the exit value of the program (0
	on success, something else on failure).  Thus this may be
	difficult to support if there are Kerberos problems.  The
	client code does capture basic problems, like the user's TGT
	having expired.

	Arguments: docbase name (required)
	Returns: The session identifier on success.  Nothing on failure.


intro-db-dctm-11.pdf
====================

    Tutorial explaining all the features and demonstrating several uses of
    Db::Documentum.


etc/dm_check_password.pl
========================

	This Perl version of dm_check_password knows how to authenticate
	users in the following ways:

	1.  Local passwords in /etc/passwd (/etc/shadow, whatever).
	2.  Against a Kerberos KDC using the user's kerberos password.
	3.  Against a Kerberos KDC using the user's kerberos service ticket.

	It should be installed setuid root with appropriate permissions.  See
	the existing dm_check_password for the right settings.

	Method #2 should be deprecated, given the fact that there doesn't
	appear to be any session encryption between the documentum client
	and server during the authentication phase.  Exposing your
	kerberos password to the network is potentially hazardous to the
	health of your network, and certain to make your network administrator
	unhappy (or it should).

	Checking works like this:

	- If the user supplies the additional optional arguments to connect,
	  the first one is ignored, and the second one is assumed to be
	  the encrypted nonce for this session, as long as it is at least
	  8 characters long.  We ignore the first one since the server
	  doesn't appear to pass it properly.

	- If a nonce is found, method 3 is tried (Kerberos service ticket).

	- If a nonce is not found, the user is authenticated against the
	  local passwd file.  If the user has a local entry, but the password
	  does not match, the session is denied.

	- If the user does not have a local passwd entry, then they are

CHANGES
Revision history for Perl extension Db::Documentum.

0.01  Mon Feb 16 11:56:16 1998 EST
	- original version; created by h2xs 1.1.1.1
0.9   Mon Feb 16 14:49:00 1998 EST
	- Initial public release version; created by briansp.
0.91  Thu Feb 19 13:00:00 1998 EST
	- Renamed to Db::Documentum at request of PAUSE folks.
0.92  Web Apr 01 15:30:00 1998 EST
	- Added prototyping behavior to XSUB definitions.
	  Added Db::Documentum::Tools.pm module.
	  Included dm_check_password.pl with distribution.
1.0   Mon Jan 25 14:54:55 EST 1999
	- Some changes to the Tools.pm module.  Added username
	  optional argument to dm_KrbConnect to satisfy my own
	  needs for shared, anonymous accounts.
	  Added corresponding code to dm_check_password.pl on server
	  side to handle shared accounts.
	  Releasing as version 1.0 just so it doesn't look so beta.
1.01  (same day) :-)
	- Made test.pl better.  Fixed MANIFEST problem (I renamed CHANGES).
	  It helps to test BEFORE you upload. :-)
1.1   Feb 18, 1999
	- Added hooks and fixes for easy building under NT.  Added nicer
	  non-Solaris OS handling in general to Makefile.PL.
   	- Added dmAPIInit() and dmAPIDeinit() calls in module to automate
   	  this.  This will require programs that have these functions in
	  them to remove them, as they're not necessary.
   	- Cleaned up test suite and simplified it.
1.2   Oct 13, 1999
	- Added dm_CreateType and dm_CreateObject functions to Tools.pm.
	- Further improvements to Makefile.PL.
	- Added tools_test.pl to test Tools.pm functionality.
1.3   Feb 2000
	- Added support for APIs with 4i -- no DFC support yet.
	- Modified the test.pl script to include tests for the Tools.pm module.
	- Changed the format of the %ATTRS hash from HoL to just a hash on the
	  dm_CreateObject sub.
1.4   September 2000
  	- Improved makefile.pl
   	- Included idql.pl in /etc.
   	- Added dm_CreatePath to Tools.pm
   	- Added dm_LocateServer to test.pl
   	- Created HTML versions of POD
1.5   February 2001
	- Added "version" subroutine to Documentum.pm. Meant to be invoked at the
	  command line with:  perl -MDb::Documentum -e Db::Documentum::version
	- Added tickler.pl in /etc.
	- Added intro-db-dctm-1-1.pdf in /etc.
1.51  October 2001
    - Updated test.pl to include version
    - Updated idql.pl script
    - Updated PPM instllation
    - Included intro-db-dctm-1-2.pdf
1.52  January 2002
    - Fixed "bug" in dm_Connect function that produced warning: 'Use of
      uninitialized value in concatenation (.) at...' on Solaris.  Thanks
      Mark Faine for identifying it.
1.53  February  2002
    - Improved Makefile.PL to better handle windows directories containing
      spaces (thanks Erik Mayo), verify the existence of key libraies, and
      improved error checking.
1.54  July 2003
    - Corrected fatal logic error in dm_CreatePath sub that didn't create
      cabinets when necessary.
    - Added Perl version to "version" subroutine in Documentum.pm.
    - Added installation data for HP-UX in Makefile.pl (thanks Jonas Linden
      for providing this info).
    - Tested with DMCL 5.2
1.6   September 2004
    - Added dm_Copy subroutine to Tools module.  dm_Copy will copy a single
      object, a folder of objects, or deep copy a folder structure.
    - Added dm_Move subroutine to Tools module.  dm_Move will move a single
      object, a folder of objects, or deep move a folder structure.
    - Added dm_Delete subroutine to Tools module. dm_Delete will delete a single
      object, a folder of objects, or deep delete a folder structure.
    - Updated test.pl script to include dm_Copy, dm_Move, dm_Delete.
    - Tested with DMCL 5.2.5 SP2.
1.61 - January 2006
    - Recompiled for Documentum 5.3 and Perl 5.8.7
1.62 - January 2006
    - Update all occurrences of email to point dm-book.com
1.63 - April 2006
    - Corrected packaging error in 1.62 that prevented it from working!



Documentum/Tools.pm
package Db::Documentum::Tools;

# Tools.pm
# (c) 2006 M. Scott Roth

use Carp;
use Exporter;
use Socket;
use Sys::Hostname;
use Db::Documentum qw(:all);
##require 5.004;

@ISA = qw(Exporter);
@EXPORT = qw();

$VERSION = '1.62';
$error = "";

@EXPORT_OK = qw(
	dm_Connect
	dm_KrbConnect
	dm_LastError
	dm_LocateServer
	dm_Locate_Child
	dm_CreateType
	dm_CreateObject
	dm_CreatePath
	dm_Copy
	dm_Delete
	dm_Move
	all
	ALL
);

%EXPORT_TAGS = (
	ALL => [qw( dm_Connect dm_KrbConnect dm_LastError dm_LocateServer
				dm_Locate_Child dm_CreateType dm_CreateObject dm_CreatePath
				dm_Copy dm_Delete dm_Move)],
	all => [qw( dm_Connect dm_KrbConnect dm_LastError dm_LocateServer
				dm_Locate_Child dm_CreateType dm_CreateObject dm_CreatePath
				dm_Copy dm_Delete dm_Move)]
);

# ---------------------------------------------------------------------------
# Defaults used by dm_CreateObject
# ---------------------------------------------------------------------------
$Delimiter = '::';
$TRUE = 1;
$FALSE = 0;

# ---------------------------------------------------------------------------
# Connects to the given docbase with the given parameters, and
# returns a session identifer.
#
#   Example: $session = dm_Connect("docbase","user","password");
#
# ---------------------------------------------------------------------------
sub dm_Connect($$$;$$) {
   my $docbase = shift;
   my $username = shift;
   my $password = shift;
   undef my $session;

   unless ($OS) {
      unless ($OS = $^O) {
	      require Config;
	      $OS = $Config::Config{'osname'};
      }
   }

   if ($OS =~ /Win/i) {
      $session = dmAPIGet("connect,$docbase,$username,$password");
   } else {
	   my $user_arg_1 = shift;
	   my $user_arg_2 = shift;
	   $session = dmAPIGet("connect,$docbase,$username,$password,$user_arg_1,$user_arg_2");
	}
	return $session;
}

# ---------------------------------------------------------------------------
# Returns documentum error information.
#
#   Example: print dm_LastError($session,3);
#
# ---------------------------------------------------------------------------
sub dm_LastError (;$$$) {
	my($session,$level,$number) = @_;
	my($return_data);
	$session = 'apisession' unless ($session);
	$level = '3' unless ($level);	# Set a default level to report.
	$number = 'all' unless ($number);
	my($message_text) = dmAPIGet("getmessage,$session,$level");
	if ($number eq "all") {
		$return_data = $message_text;
	} else {
		my(@message_list) = split('\n',$message_text);
		for ($i = 0 ; $i < $number ; $i++) {
			$return_data .= sprintf("%s\n",$message_list[$i]);
		}
	}
	$return_data;
}

# ---------------------------------------------------------------------------
# Create a Documentum object and populate attributes.
#
#  Example:
#     %ATTRS = (object_name =>  'test_doc2',
#               title       =>  'My Test Doc 2',
#               authors     =>  'Scott 1::Scott 2',
#               keywords    =>  'Scott::Test::Doc::2',
#               r_version_label => 'TEST');
#
#     $doc_id = dm_CreateObject ("dm_document",%ATTRS);
#
# ---------------------------------------------------------------------------
sub dm_CreateObject($;%) {

   my $dm_type = shift;
   my %attrs = @_;
   my $api_stat = 1;

   my $obj_id = dmAPIGet("create,c,$dm_type");

   if ($obj_id) {
      foreach my $attr (keys %attrs) {
         if (dmAPIGet("repeating,c,$obj_id,$attr")) {
            my @r_attr = split($Db::Documentum::Tools::Delimiter,$attrs{$attr});
            foreach (@r_attr) {
                $api_stat = 0 unless dmAPISet("append,c,$obj_id,$attr",$_);
            }
         }
         else {
            $api_stat = 0 unless dmAPISet("set,c,$obj_id,$attr",$attrs{$attr});
         }
      }
   } else {
    $api_stat = 0;
   }

   return (! $obj_id || ! $api_stat) ? undef : $obj_id;
}

# ---------------------------------------------------------------------------
# Create a new Documentum object type.
#
#  Example:
#     %field_defs = (cat_id    => 'char(16)',
#                    loc       => 'char(64)',
#                    editions  => 'char(6) REPEATING');
#     $rv = dm_CreateType ("my_document","dm_document",%field_defs);
#
# ---------------------------------------------------------------------------
sub dm_CreateType($$;%) {

   my $name = shift;
   my $super_type = shift;
   my %field_defs = @_;
   my $sql_body = "";

   if (keys %field_defs) {
      foreach my $field (keys %field_defs) {
         $sql_body .= "$field $field_defs{$field},";
      }
      $sql_body =~ s/\,$//;
      $sql_body = "($sql_body)";
   }

   my $sql = "CREATE TYPE $name $sql_body WITH SUPERTYPE $super_type";
   my $api_stat =  dmAPIExec("execquery,c,,$sql");

   if ($api_stat) {
      my $col_id = dmAPIGet("getlastcoll,c,");
      dmAPIExec("close,c,$col_id") if $col_id;
   }

   return $api_stat;
}

# ---------------------------------------------------------------------------
# Create a new folder structure in the Docbase.
#
#  Example:   $path = dm_CreatePath ('/Temp/Test/Unit-1');
#
# ---------------------------------------------------------------------------
sub dm_CreatePath($) {
    my $path = shift;

    # Break path into heirarchical elements
    my @dirs = split("/",$path);

    my $dm_path = "";

    # if it already exists, just return
    my $dir_id = dmAPIGet("id,c,dm_folder where any r_folder_path = \'$path\'");
    return $dir_id if $dir_id;

    # Test each heirarchical path for existance
    foreach my $dir (@dirs) {
        if ($dir =~ /^\w/) {
            $dm_path .= "/$dir";

            # Does this partial path exist?
            $dir_id = dmAPIGet("id,c,dm_folder where any r_folder_path = \'$dm_path\'");

            # If not, create it
            if (! $dir_id) {
                my %dir_attrs = ();
                $dir_attrs{'object_name'} = $dir;

                # Create the cabinet if needed
                if ($dm_path eq "/$dir") {
                    $dir_id = dm_CreateObject("dm_cabinet",%dir_attrs);
                    return undef unless dmAPIExec("save,c,$dir_id");
                # Create a folder
                } else {
                    $dir_id = dm_CreateObject("dm_folder",%dir_attrs);
                    return undef unless $dir_id;

                    # Link it to its parent
                    my $folder_path = $dm_path;
                    $folder_path =~ s/\/$dir$//;
                    if ($folder_path =~ /\w+/) {
                        return undef
                            unless dmAPIExec("link,c,$dir_id,\'$folder_path\'");
                        return undef
                            unless dmAPIExec("save,c,$dir_id");
                    }
                }
            }
        }
    }
    return $dir_id;
}

# ---------------------------------------------------------------------------
# Find the active server for a given docbase.
#
#   Example:  $server = dm_LocateServer($docbase);
#
# ---------------------------------------------------------------------------
sub dm_LocateServer ($) {
	my $docbase = shift;
	my $locator = dmAPIGet("getservermap,apisession,$docbase");
    my $hostname = dmAPIGet("get,apisession,$locator,i_host_name")
        if ($locator);

    return ($hostname) ? $hostname : undef;
}

# ---------------------------------------------------------------------------
# Returns the object id (if any) of the object to which this object is
# a parent based on the relation type.
#
#  Example:  $child = dm_Locate_Child($session,$object_id,$relation);
#
# ---------------------------------------------------------------------------
sub dm_Locate_Child ($$$) {
	my($ss,$object_id,$relation_type) = @_;

	my($relation_obj_id) = dmAPIGet("id,$ss,dm_relation where parent_id = '$object_id' and relation_name = '$relation_type'");
	if (! $relation_obj_id) {
	    return $FALSE;
	}

	my($child_object_id) = dmAPIGet("get,$ss,$relation_obj_id,child_id");
	if (! $child_object_id) {
	     return $FALSE;
	}

	return $child_object_id;
}

# ---------------------------------------------------------------------------
# Copies an object and optionally moves it to a new location specified by
# $to_folder.  If $to_folder is undefined, the copy will reside with the
# original.  Do not copy folders and cabinets with this routine.  Returns the
# r_object_id of copy.
#
#  Example:  $object_id = dm_Copy($orig_obj_id, $to_folder);
#
#  Note:  $to_folder can be the r_object_id of the destination folder,
#         or the full path (e.g., "/Temp/2004/Q1/Jan/15").  If the path
#         does not exist, it is created by calling dm_CreatePath.
#
# ---------------------------------------------------------------------------
sub dm_Copy($;$) {
    my $orig_obj_id = shift;
    my $to_folder = shift;

    my $obj_id = dmAPIGet("saveasnew,c,$orig_obj_id");
    return undef unless $obj_id;

    if (defined $to_folder) {
        dm_LastError("c","3","all") unless dm_Move($obj_id,$to_folder);
    }

    return $obj_id;
}

# ---------------------------------------------------------------------------
# Moves an object from the location specified by $from_folder, to the location
# specified by $to_folder.  If $from_folder is undefined, unlink from
# all locations.  Returns true for success, false for failure.
#
#  Example:  $rv = dm_Move($object_id, $to_folder, $from_folder);
#
#  Note:  $to_folder can be the r_object_id of the destination folder,
#         or the full path (e.g., "/Temp/2004/Q1/Jan/15").  If the path
#         does not exist, it is created by calling dm_CreatePath.
#
#  Note:  $from_folder can be the r_object_id of the current folder or its
#         full path (e.g., "/Temp/2004/Q2/Jun/1").
#
# ---------------------------------------------------------------------------
sub dm_Move($$;$) {
    my $obj = shift;
    my $to_folder = shift;
    my $from_folder = shift;

    # if to_folder is not an object id it must be a path
    # use dm_CreatePath to get object id of folder
    if ($to_folder !~ /\w{16}/) {
        $to_folder = dm_CreatePath($to_folder);
    }

    # if to_folder doesn't point to a folder or cabinet, abort move operation
    if ($to_folder !~ /^0[bc]/i) {
        return $FALSE;
    }

    # check that $from_folder is valid if it is defined
    if ($from_folder) {
        if ($from_folder !~ /\w{16}/) {

            # if not an id, assume from_folder is a path.  Get id of folder
            if ($from_folder !~ /^\//) {
                $from_folder = "\/$from_folder";
            }
            $from_folder = dmAPIGet("id,c,dm_folder where any r_folder_path = \'$from_folder\'");
        }

        # if from_folder doesn't point to a folder or cabinet, abort move operation
        if ($from_folder !~ /^0[bc]/i) {
            return $FALSE;
        }
    }

    # link object to new location
    return $FALSE unless dmAPIExec("link,c,$obj,$to_folder");
    return $FALSE unless dmAPIExec("save,c,$obj");

    # unlink object from previous location
    if (defined $from_folder) {
        return $FALSE unless dmAPIExec("unlink,c,$obj,$from_folder");
        return $FALSE unless dmAPIExec("save,c,$obj");
    } else {

        # unlink from all previous locations
        my $cnt = dmAPIGet("values,c,$obj,i_folder_id");

        for (my $i = 0; $i < $cnt; $i ++) {
            my $old_folder_id = dmAPIGet("get,c,$obj,i_folder_id[$i]");

            if (defined $old_folder_id) {

                # don't unlink from the link we just made
                if ($old_folder_id ne $to_folder) {
                    return $FALSE unless dmAPIExec("unlink,c,$obj,$old_folder_id");
                    return $FALSE unless dmAPIExec("save,c,$obj");
                }
            }
        }
    }

    return $TRUE;
}

# ---------------------------------------------------------------------------
# Deletes the object specified by $obj_id.  If the optional $all argument is
# set to true, will delete all versions of the object specified by $obj_id.
# By default, $all is false.  If the object specified by $obj_id is a cabinet
# or folder, this sub will execute a deep delete and remove all versions of
# all objects contained in the cabinet or folder.  Returns true on success,
# false on failure.
#
#  Example:  $rv = dm_Delete($obj_id, $all);
#
# ---------------------------------------------------------------------------
sub dm_Delete($;$) {
    my $obj_id = shift;
    my $all = shift;
    $all = $FALSE unless ($all);

    # if not a cabinet or folder, just destroy object
    if ($obj_id !~ /^0[bc]/) {
        if ($all) {

            # destroy all versions using DQL--it's easier
            my $obj_chron_id = dmAPIGet("get,c,$obj_id,i_chronicle_id");
            if ($obj_chron_id) {
                my $query = "delete dm_sysobject (all) objects where i_chronicle_id = '$obj_chron_id'";
                my $col = dmAPIGet("query,c,$query");

                # get delete results
                while (dmAPIExec("next,c,$col")) {
                    return $FALSE unless dmAPIGet("get,c,$col,objects_deleted");
                }
                dmAPIExec("close,c,$col");
            }
        } else {
            return $FALSE unless dmAPIExec("destroy,c,$obj_id");
        }
    } else {
        # if obj_id is folder or cabinet do deep delete
        my $query = "select r_object_id from dm_sysobject (all) where folder(id('$obj_id'))";
        my $col = dmAPIGet("query,c,$query");
        my @del_tree = ();

        # build array with all objects returned in query
        while (dmAPIExec("next,c,$col")) {
            my $del_obj_id = dmAPIGet("get,c,$col,r_object_id");
            push(@del_tree,$del_obj_id);
        }
        my $rv = dmAPIExec("close,c,$col");

        # process array and recursively call dm_Delete to delete object
        foreach my $id (@del_tree) {
            return $FALSE unless dm_Delete($id,$all);
        }

        # once the root folder is empty, delete it
        return $FALSE unless dmAPIExec("destroy,c,$obj_id");
    }
    return $TRUE;
}

# ---------------------------------------------------------------------------
# !! EXPERIMENTAL !!
# dm_KrbConnect - Obtains a Documentum client session using a K4 session
# 	ticket.  Requires a compatible dm_check_password utility
#   on the server side.
#
#  Example:  $session = dm_KrbConnect($docbase,$username);
#
# ---------------------------------------------------------------------------
sub dm_KrbConnect ($;$) {
	my($docbase,$username) = @_;
	my($service) = 'documentum';
	my($time) = time();
	my($nonce_prefix) = "KERBEROS_V4_NONCE__";
	my($nonce_data) = "${nonce_prefix}${time}";

	# Find the documentum server we're going to be connecting to from
	# whatever docbroker we're going to find.
	my($server_hostname) = dm_LocateServer($docbase);
	if (! $server_hostname) {
		# dm_LocateServer sets Documentum::Tools::error for us.
		return;
	}

	# We need the address as a four byte packed string for krb_mk_priv.
	my($server_inaddr) = inet_aton($server_hostname);
	if(! $server_inaddr) {
		${'error'} = "Unable to obtain server address.\n";
	}

	my($client_hostname) = hostname();

	if (! $client_hostname) {
		${'error'} = "Unable to obtain local hostname.";
		return;
	}

	# We need the address as a four byte packed string for krb_mk_priv.
	my($client_inaddr) = inet_aton($client_hostname);
	if(! $client_inaddr) {
		${'error'} = "Unable to obtain local address.\n";
	}

	my($realm) = Krb4::realmofhost($server_hostname);
	if (! $realm) {
		${'error'} = "Unable to determine realm of host $server_hostname: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}
	my($phost) = Krb4::get_phost($server_hostname,$realm,$service);
	if (! $phost) {
		${'error'} = "Unable to determine instance for host $server_hostname in realm $realm: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}
	my($ticket) = Krb4::mk_req($service,$phost,$realm,200);
	if (! $ticket) {
		${'error'} = "Unable to obtain a ticket: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}
	my($ticket_data) = $ticket->dat;

	my($creds) = Krb4::get_cred($service,$phost,$realm);
	if (! $creds) {
		${'error'} = "Unable to obtain credential data: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}
	# If the caller didn't specify a username to log in as, then
	# use the one in their ticket (i.e. theirs).
	if (! $username) {
		$username = $creds->pname;
	}
	my($session_key) = $creds->session;
	my($key_schedule) = Krb4::get_key_sched($session_key);

	if (! $key_schedule) {
		${'error'} = "Unable to obtain encryption key schedule: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}

	# Construct a nonce for this session.  Here we will use the time
	# encrypted with the session key.
	my($nonce) = Krb4::mk_priv($nonce_data,$key_schedule,$session_key,
									$client_inaddr,$server_inaddr);

#	print STDERR "session_key: $session_key\n";
#	print STDERR "time: $time\n";
#	print STDERR "key_schedule: $key_schedule\n";
#	print STDERR "client_inaddr: $client_inaddr\n";
#	print STDERR "client_addr:", inet_aton($client_inaddr), "\n";
#	print STDERR "server_inaddr: $server_inaddr\n";
#	print STDERR "server_addr:", inet_aton($server_inaddr), "\n";
#	print STDERR "nonce_data: $nonce_data\n";
#	print STDERR "nonce: $nonce\n";

	if (! $nonce) {
		${'error'} = "Unable to obtain encrypt nonce: ";
		${'error'} .= Krb4::get_err_txt($Krb4::error);
		${'error'} .= " [$Krb4::error]\n";
		return;
	}
#	print STDERR Krb4::get_err_txt($Krb4::error), "\n";

	# uuencode ticket data, then encode it with URI-style encoding
	my($ticket_data_encoded) = pack "u", $ticket_data;
	$ticket_data_encoded =~ s/([^A-Za-z0-9])/uc sprintf("%%%02x",ord($1))/eg;

	# Same thing for nonce.
	my($nonce_encoded) = pack "u", $nonce;
	$nonce_encoded =~ s/([^A-Za-z0-9])/uc sprintf("%%%02x",ord($1))/eg;

#	print "nonce_encoded: $nonce_encoded\n";

	# Okay.  Now we've got an encoded service ticket for this session.
	# Send it as our password to the documentum server for
	# validation.  We include the nonce as both additional
	# params, because connect doesn't seem the pass the first one properly.
	my($session_id) = dmAPIGet("connect,$docbase,$username,$ticket_data_encoded,,$nonce_encoded");
	if (! $session_id) {
		${'error'} = "Unable to obtain a docbase session id:\n";
		${'error'} .= dm_LastError();
		return;
	} else {
		return $session_id;
	}
}

## -----------------
##      <SDG><
## -----------------

1;
__END__

=head1 NAME

Db::Documentum::Tools - Support functions for Db::Documentum.

=head1 SYNOPSIS

    use Db::Documentum::Tools qw(:all);

    $session_id = dm_Connect($docbase,$user,$password);
    $session_id = dm_Connect($docbase,$user,$password,
                             $user_arg_1,$user_arg_2);

    $error_msg = dm_LastError();
    $error_msg = dm_LastError($session_id);
    $error_msg = dm_LastError($session_id,1);
    $error_msg = dm_LastError($session_id,$level,$number);

    $object_id = dm_CreateObject("dm_document",%ATTRS);
    $object_id = dm_CreateObject("dm_document");

    $api_stat = dm_CreateType("my_document","dm_document",%field_defs);
    $api_stat = dm_CreateType("my_document","dm_document");

    $obj_id = dm_CreatePath('/Temp/Test/Unit-1');

    $hostname = dm_LocateServer($docbase);

    $child = dm_Locate_Child($session,$object_id,$relation);

    $object_id = dm_Copy($orig_obj_id, $to_folder);
    $object_id = dm_Copy($orig_obj_id, '/Temp/Test/Unit-1');

    $rv = dm_Move($object_id, $to_folder);
    $rv = dm_Move($object_id, '/Temp/Test/Unit-2');
    $rv = dm_Move($object_id, $to_folder, $from_folder);
    $rv = dm_Move($object_id, $to_folder, '/Temp/Test/Unit-1');
    $rv = dm_Move($object_id, '/Temp/Test/Unit-2', $from_folder);
    $rv = dm_Move($object_id, '/Temp/Test/Unit-2', '/Temp/Test/Unit-1');

    $rv = dm_Delete($obj_id, $all);
    $rv = dm_Delete($obj_id, 1);

    See scripts in /etc for more examples.

=head1 DESCRIPTION

Db::Documentum::Tools is a collection of frequently used Documentum procedures
encapsulated by Perl.

=head1 LICENSE

The Documentum perl extension may be redistributed under the same terms as Perl.
The Documentum EDMS is a commercial product.  The product name, concepts,
and even the mere thought of the product are the sole property of
Documentum, Inc. and its shareholders.

=head1 AUTHOR

M. Scott Roth, C<scott@dm-book.com>

=head1 SEE ALSO

Db::Documentum.

=cut
